### PR TITLE
Add Azure ARM discovery plugin

### LINKED ISSUE #19146: Add support for Azure Resource Manager (ARM)
As explained in #19144 we could create a new plugin named `discovery-azure-arm` which is based on the [newest versions of the Azure SDK](https://github.com/Azure/azure-sdk-for-java/wiki/Azure-SDK-for-Java-Features).

Classic was based on private key / password which was bad for what we want to achieve with such a plugin.
ARM uses now key / secrets which are really much more flexible and safer.

Basically you can now write code like [this example](https://github.com/Azure/azure-sdk-for-java/wiki/Getting-Started#client-creation):

``` java
ComputeManagementClient client;

// Default environment, no credentials
client = new ComputeManagementClientImpl();

// Custom environment with AzureChinaCloud endpoint
String chinaCloudEndpoint = "https://management.chinacloudapi.cn/";
client = new ComputeManagementClientImpl(chinaCloudEndpoint);

// Default environment with credentials
ApplicationTokenCredentials credentials = new ApplicationTokenCredentials(
        "client-id", "tenant-id", "secret", null);
client = new ComputeManagementClientImpl(credentials);

// Custom environment with credentials
client = new ComputeManagementClientImpl(chinaCloudEndpoint, credentials);

// Fully customized, with HTTP client and Retrofit modifiable
client = new ComputeManagementClientImpl(
        chinaCloudEndpoint,
        credentials,
        new OkHttpClient.Builder(),
        new Retrofit.Builder());
```

### PR DESCRIPTION
**Pull request by dadoonet**
Supported settings so far:

```yml
cloud:
    azure-arm:
        client_id: FILL_WITH_YOUR_CLIENT_ID
        secret: FILL_WITH_YOUR_SECRET
        tenant_id: FILL_WITH_YOUR_TENANT
        subscription_id: FILL_WITH_YOUR_SUBSCRIPTION_ID

discovery:
    zen.hosts_provider: azure-arm
    azure-arm:
        host:
            type: private_ip
            name: azure-esnode-master-*
            group_name: azure-preprod
            region: westeurope
        refresh_interval: 10s
```

Closes #19146

### CODE REVIEW

File: plugins/discovery-azure-arm/src/main/java/org/elasticsearch/cloud/azure/arm/AzureManagementService.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        /**
+         * Azure ARM client Id
+         */
+        public static final Setting<String> CLIENT_ID_SETTING =
+            Setting.simpleString("cloud.azure-arm.client_id", Property.NodeScope, Property.Filtered);
+        /**
+         * Azure ARM secret
+         */
+        public static final Setting<String> SECRET_SETTING =
+            Setting.simpleString("cloud.azure-arm.secret", Property.NodeScope, Property.Filtered);
```
Comment (rjernst) [inline plugins/discovery-azure-arm/src/main/java/org/elasticsearch/cloud/azure/arm/AzureManagementService.java]: I don't think we should be adding new secret settings that are plain strings.

Comment (dadoonet) [inline plugins/discovery-azure-arm/src/main/java/org/elasticsearch/cloud/azure/arm/AzureManagementService.java]: Agreed.

File: docs/plugins/discovery-azure-arm.asciidoc [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+    Either `public_ip` or `private_ip` (default). Azure discovery will use the
+    one you set to ping other nodes.
+
+`discovery.azure-arm.host.name`::
+
+    You can filter virtual machines you would like to connect to by entering a name here. It can be a wildcard
+    like `azure-esnode-*`.
+
+`discovery.azure-arm.host.group_name`::
```
Comment (Mpdreamz) [inline docs/plugins/discovery-azure-arm.asciidoc]: Can we rename this to `discovery.azure-arm.resource-group`, `group` by itself is a tad ambiguous and `resource group` is a well understood azure concept

Comment (dadoonet) [inline docs/plugins/discovery-azure-arm.asciidoc]: Sure! Will do!

**Comment by dadoonet** [comment]
@Mpdreamz I updated the repo with new changes:

* Fixed the missing lib issue
* Renamed `group_name` to `resource_group`
* Fixed an issue when using a wildcard in `resource_group`

**Comment by Mpdreamz** [comment]
Now getting security manager exceptions @dadoonet 

> [2017-01-20T12:52:57,233][WARN ][o.e.d.z.ZenDiscovery     ] [data-0] Ping execution failed

> │java.util.concurrent.ExecutionException: java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "accessDeclaredMembers")

**Comment by jasontedor** [comment]
We need the entire stack trace here.

**Comment by dadoonet** [comment]
@Mpdreamz Interesting. Can you start elasticsearch with this option `-Djava.security.debug="access,failure"`?

ES_JAVA_OPTS="-Djava.security.debug='access,failure'" ./bin/elasticsearch

**Comment by Mpdreamz** [comment]
<details><summary>My bad @jasontedor: click to expand</summary><p>

>[2017-01-20T12:52:57,233][WARN ][o.e.d.z.ZenDiscovery     ] [data-0] Ping execution failed
java.util.concurrent.ExecutionException: java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "accessDeclaredMembers")
        at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357) ~[?:1.8.0_121]
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895) ~[?:1.8.0_121]
        at org.elasticsearch.discovery.zen.ZenDiscovery.pingAndWait(ZenDiscovery.java:1000) [elasticsearch-6.0.0-alpha1-SNAPSHOT.jar:6.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.discovery.zen.ZenDiscovery.findMaster(ZenDiscovery.java:860) [elasticsearch-6.0.0-alpha1-SNAPSHOT.jar:6.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.discovery.zen.ZenDiscovery.innerJoinCluster(ZenDiscovery.java:372) [elasticsearch-6.0.0-alpha1-SNAPSHOT.jar:6.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.discovery.zen.ZenDiscovery.access$3800(ZenDiscovery.java:80) [elasticsearch-6.0.0-alpha1-SNAPSHOT.jar:6.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.discovery.zen.ZenDiscovery$JoinThreadControl$1.run(ZenDiscovery.java:1176) [elasticsearch-6.0.0-alpha1-SNAPSHOT.jar:6.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:527) [elasticsearch-6.0.0-alpha1-SNAPSHOT.jar:6.0.0-alpha1-SNAPSHOT]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_121]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_121]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_121]
Caused by: java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "accessDeclaredMembers")
        at java.security.AccessControlContext.checkPermission(AccessControlContext.java:472) ~[?:1.8.0_121]
        at java.security.AccessController.checkPermission(AccessController.java:884) ~[?:1.8.0_121]
        at java.lang.SecurityManager.checkPermission(SecurityManager.java:549) ~[?:1.8.0_121]
        at java.lang.Class.checkMemberAccess(Class.java:2348) ~[?:1.8.0_121]
        at java.lang.Class.getEnclosingMethod(Class.java:1037) ~[?:1.8.0_121]
        at sun.reflect.generics.scope.ClassScope.computeEnclosingScope(ClassScope.java:50) ~[?:?]
        at sun.reflect.generics.scope.AbstractScope.getEnclosingScope(AbstractScope.java:78) ~[?:?]
        at sun.reflect.generics.scope.AbstractScope.lookup(AbstractScope.java:96) ~[?:?]
        at sun.reflect.generics.factory.CoreReflectionFactory.findTypeVariable(CoreReflectionFactory.java:110) ~[?:?]
        at sun.reflect.generics.visitor.Reifier.visitTypeVariableSignature(Reifier.java:165) ~[?:?]
        at sun.reflect.generics.tree.TypeVariableSignature.accept(TypeVariableSignature.java:43) ~[?:?]
        at sun.reflect.generics.visitor.Reifier.reifyTypeArguments(Reifier.java:68) ~[?:?]
        at sun.reflect.generics.visitor.Reifier.visitClassTypeSignature(Reifier.java:138) ~[?:?]
        at sun.reflect.generics.tree.ClassTypeSignature.accept(ClassTypeSignature.java:49) ~[?:?]
        at sun.reflect.generics.repository.ClassRepository.getSuperclass(ClassRepository.java:90) ~[?:?]
        at java.lang.Class.getGenericSuperclass(Class.java:777) ~[?:1.8.0_121]
        at com.fasterxml.jackson.core.type.TypeReference.<init>(TypeReference.java:33) ~[jackson-core-2.8.6.jar:2.8.6]
        at com.microsoft.rest.serializer.JacksonMapperAdapter$1.<init>(JacksonMapperAdapter.java:179) ~[?:?]
        at com.microsoft.rest.serializer.JacksonMapperAdapter.deserialize(JacksonMapperAdapter.java:179) ~[?:?]
        at com.microsoft.rest.ServiceResponseBuilder.buildBody(ServiceResponseBuilder.java:289) ~[?:?]
        at com.microsoft.rest.ServiceResponseBuilder.build(ServiceResponseBuilder.java:141) ~[?:?]
        at com.microsoft.azure.management.compute.implementation.VirtualMachinesInner.listDelegate(VirtualMachinesInner.java:1101) ~[?:?]
        at com.microsoft.azure.management.compute.implementation.VirtualMachinesInner.access$600(VirtualMachinesInner.java:46) ~[?:?]
        at com.microsoft.azure.management.compute.implementation.VirtualMachinesInner$36.call(VirtualMachinesInner.java:1091) ~[?:?]
        at com.microsoft.azure.management.compute.implementation.VirtualMachinesInner$36.call(VirtualMachinesInner.java:1087) ~[?:?]
        at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:69) ~[?:?]
        at retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$RequestArbiter.request(RxJavaCallAdapterFactory.java:173) ~[?:?]
        at rx.Subscriber.setProducer(Subscriber.java:211) ~[?:?]
        at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102) ~[?:?]
        at retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$CallOnSubscribe.call(RxJavaCallAdapterFactory.java:152) ~[?:?]
        at retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$CallOnSubscribe.call(RxJavaCallAdapterFactory.java:138) ~[?:?]
        at rx.Observable.unsafeSubscribe(Observable.java:9861) ~[?:?]
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48) ~[?:?]
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33) ~[?:?]
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48) ~[?:?]
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) ~[?:?]
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48) ~[?:?]
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) ~[?:?]
        at rx.Observable.subscribe(Observable.java:9957) ~[?:?]
        at rx.Observable.subscribe(Observable.java:9924) ~[?:?]
        at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:445) ~[?:?]
        at rx.observables.BlockingObservable.single(BlockingObservable.java:342) ~[?:?]
        at com.microsoft.azure.management.compute.implementation.VirtualMachinesInner.list(VirtualMachinesInner.java:1006) ~[?:?]
        at com.microsoft.azure.management.compute.implementation.VirtualMachinesImpl.listByGroup(VirtualMachinesImpl.java:68) ~[?:?]
        at org.elasticsearch.cloud.azure.arm.AzureManagementServiceImpl.lambda$getVirtualMachines$1(AzureManagementServiceImpl.java:103) ~[?:?]
        at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_121]
        at org.elasticsearch.cloud.azure.arm.AzureManagementServiceImpl.getVirtualMachines(AzureManagementServiceImpl.java:92) ~[?:?]
        at org.elasticsearch.discovery.azure.arm.AzureArmUnicastHostsProvider.buildDynamicNodes(AzureArmUnicastHostsProvider.java:100) ~[?:?]
        at org.elasticsearch.discovery.zen.UnicastZenPing.ping(UnicastZenPing.java:302) ~[elasticsearch-6.0.0-alpha1-SNAPSHOT.jar:6.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.discovery.zen.UnicastZenPing.ping(UnicastZenPing.java:279) ~[elasticsearch-6.0.0-alpha1-SNAPSHOT.jar:6.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.discovery.zen.ZenDiscovery.pingAndWait(ZenDiscovery.java:993) ~[elasticsearch-6.0.0-alpha1-SNAPSHOT.jar:6.0.0-alpha1-SNAPSHOT]
        ... 8 more
</p></details>

**Comment by dadoonet** [comment]
I opened https://github.com/Azure/autorest-clientruntime-for-java/issues/136 on azure side.
The problem here is that we load jackson with the core security manager which does not allow `java.lang.RuntimePermission "accessDeclaredMembers"` so even if I change the policy for the plugin, when Azure SDK will call Jackson, it will call it with the core security manager policy.

Let see what Azure team can do to help us in that context.

If they can't do anything, I believe that we would have to shade somehow a version of Azure SDK and Jackson within a single flat jar where we can relocate then Jackson classes... I'm not a fan of this TBH.

**Comment by rjernst** [comment]
That looks like an issue in jackson itself.  The `TypeReference` ctor calls `Class.getGenericSuperclass()`, which from an anonymous class requires that permission (and the adaptor in azure creates an anonymous `TypeReference` instance.  It seems like a web of complicated design that may not be possible to untangle (even with a flat jar).

**Comment by dadoonet** [comment]
Thanks @rjernst. I opened https://github.com/FasterXML/jackson-core/issues/347

**Comment by ejsmith** [comment]
Will this support scalesets?

**Comment by hglkrijger** [comment]
@dadoonet, just wondering if this is still getting merged?

**Comment by dadoonet** [comment]
I worked again today on this.
From a local unit test, I'm able to get the list of the virtual machines but when I run it from an elasticsearch node, I'm still getting a SM error:

```
[2017-07-12T15:44:40,411][WARN ][o.e.d.z.ZenDiscovery     ] [Hb45eoY] Ping execution failed
java.util.concurrent.ExecutionException: java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "accessDeclaredMembers")
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357) ~[?:1.8.0_131]
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895) ~[?:1.8.0_131]
	at org.elasticsearch.discovery.zen.ZenDiscovery.pingAndWait(ZenDiscovery.java:1064) [elasticsearch-6.0.0-beta1-SNAPSHOT.jar:6.0.0-beta1-SNAPSHOT]
	at org.elasticsearch.discovery.zen.ZenDiscovery.findMaster(ZenDiscovery.java:907) [elasticsearch-6.0.0-beta1-SNAPSHOT.jar:6.0.0-beta1-SNAPSHOT]
	at org.elasticsearch.discovery.zen.ZenDiscovery.innerJoinCluster(ZenDiscovery.java:433) [elasticsearch-6.0.0-beta1-SNAPSHOT.jar:6.0.0-beta1-SNAPSHOT]
	at org.elasticsearch.discovery.zen.ZenDiscovery.access$2500(ZenDiscovery.java:87) [elasticsearch-6.0.0-beta1-SNAPSHOT.jar:6.0.0-beta1-SNAPSHOT]
	at org.elasticsearch.discovery.zen.ZenDiscovery$JoinThreadControl$1.run(ZenDiscovery.java:1266) [elasticsearch-6.0.0-beta1-SNAPSHOT.jar:6.0.0-beta1-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:569) [elasticsearch-6.0.0-beta1-SNAPSHOT.jar:6.0.0-beta1-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
Caused by: java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "accessDeclaredMembers")
	at java.security.AccessControlContext.checkPermission(AccessControlContext.java:472) ~[?:1.8.0_131]
	at java.security.AccessController.checkPermission(AccessController.java:884) ~[?:1.8.0_131]
	at java.lang.SecurityManager.checkPermission(SecurityManager.java:549) ~[?:1.8.0_131]
	at java.lang.Class.checkMemberAccess(Class.java:2348) ~[?:1.8.0_131]
	at java.lang.Class.getEnclosingMethod(Class.java:1037) ~[?:1.8.0_131]
	at sun.reflect.generics.scope.ClassScope.computeEnclosingScope(ClassScope.java:50) ~[?:?]
	at sun.reflect.generics.scope.AbstractScope.getEnclosingScope(AbstractScope.java:78) ~[?:?]
	at sun.reflect.generics.scope.AbstractScope.lookup(AbstractScope.java:96) ~[?:?]
	at sun.reflect.generics.factory.CoreReflectionFactory.findTypeVariable(CoreReflectionFactory.java:110) ~[?:?]
	at sun.reflect.generics.visitor.Reifier.visitTypeVariableSignature(Reifier.java:165) ~[?:?]
	at sun.reflect.generics.tree.TypeVariableSignature.accept(TypeVariableSignature.java:43) ~[?:?]
	at sun.reflect.generics.visitor.Reifier.reifyTypeArguments(Reifier.java:68) ~[?:?]
	at sun.reflect.generics.visitor.Reifier.visitClassTypeSignature(Reifier.java:138) ~[?:?]
	at sun.reflect.generics.tree.ClassTypeSignature.accept(ClassTypeSignature.java:49) ~[?:?]
	at sun.reflect.generics.repository.ClassRepository.getSuperclass(ClassRepository.java:90) ~[?:?]
	at java.lang.Class.getGenericSuperclass(Class.java:777) ~[?:1.8.0_131]
	at com.fasterxml.jackson.core.type.TypeReference.<init>(TypeReference.java:33) ~[jackson-core-2.8.6.jar:2.8.6]
	at com.microsoft.rest.serializer.JacksonMapperAdapter$1.<init>(JacksonMapperAdapter.java:179) ~[?:?]
	at com.microsoft.rest.serializer.JacksonMapperAdapter.deserialize(JacksonMapperAdapter.java:179) ~[?:?]
	at com.microsoft.rest.ServiceResponseBuilder.buildBody(ServiceResponseBuilder.java:289) ~[?:?]
	at com.microsoft.rest.ServiceResponseBuilder.build(ServiceResponseBuilder.java:141) ~[?:?]
	at com.microsoft.azure.management.compute.implementation.VirtualMachinesInner.listAllDelegate(VirtualMachinesInner.java:1201) ~[?:?]
	at com.microsoft.azure.management.compute.implementation.VirtualMachinesInner.access$700(VirtualMachinesInner.java:46) ~[?:?]
	at com.microsoft.azure.management.compute.implementation.VirtualMachinesInner$42.call(VirtualMachinesInner.java:1191) ~[?:?]
	at com.microsoft.azure.management.compute.implementation.VirtualMachinesInner$42.call(VirtualMachinesInner.java:1187) ~[?:?]
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:69) ~[?:?]
	at retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$RequestArbiter.request(RxJavaCallAdapterFactory.java:173) ~[?:?]
	at rx.Subscriber.setProducer(Subscriber.java:211) ~[?:?]
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102) ~[?:?]
	at retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$CallOnSubscribe.call(RxJavaCallAdapterFactory.java:152) ~[?:?]
	at retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$CallOnSubscribe.call(RxJavaCallAdapterFactory.java:138) ~[?:?]
	at rx.Observable.unsafeSubscribe(Observable.java:9861) ~[?:?]
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48) ~[?:?]
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33) ~[?:?]
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48) ~[?:?]
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) ~[?:?]
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48) ~[?:?]
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) ~[?:?]
	at rx.Observable.subscribe(Observable.java:9957) ~[?:?]
	at rx.Observable.subscribe(Observable.java:9924) ~[?:?]
	at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:445) ~[?:?]
	at rx.observables.BlockingObservable.single(BlockingObservable.java:342) ~[?:?]
	at com.microsoft.azure.management.compute.implementation.VirtualMachinesInner.listAll(VirtualMachinesInner.java:1113) ~[?:?]
	at com.microsoft.azure.management.compute.implementation.VirtualMachinesImpl.list(VirtualMachinesImpl.java:63) ~[?:?]
	at org.elasticsearch.cloud.azure.arm.AzureManagementServiceImpl.lambda$getVirtualMachines$2(AzureManagementServiceImpl.java:126) ~[?:?]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_131]
	at org.elasticsearch.cloud.azure.arm.AzureManagementServiceImpl.getVirtualMachines(AzureManagementServiceImpl.java:118) ~[?:?]
	at org.elasticsearch.discovery.azure.arm.AzureArmUnicastHostsProvider.buildDynamicNodes(AzureArmUnicastHostsProvider.java:104) ~[?:?]
	at org.elasticsearch.discovery.zen.UnicastZenPing.ping(UnicastZenPing.java:309) ~[elasticsearch-6.0.0-beta1-SNAPSHOT.jar:6.0.0-beta1-SNAPSHOT]
	at org.elasticsearch.discovery.zen.UnicastZenPing.ping(UnicastZenPing.java:286) ~[elasticsearch-6.0.0-beta1-SNAPSHOT.jar:6.0.0-beta1-SNAPSHOT]
	at org.elasticsearch.discovery.zen.ZenDiscovery.pingAndWait(ZenDiscovery.java:1057) ~[elasticsearch-6.0.0-beta1-SNAPSHOT.jar:6.0.0-beta1-SNAPSHOT]
	... 8 more
```

I don't know how to solve it. If anyone has an idea, I'll be happy to hear. I'm going to update https://github.com/Azure/autorest-clientruntime-for-java/issues/136 as well.

**Comment by dakrone** [comment]
@dadoonet what's the word on this? Are you still working on it?

**Comment by rjernst** [comment]
@dadoonet Maybe this PR should be closed until it is reviewable? It seems there are still security manager issues that need to be worked out, and I expect likely also needs updating to store some settings in the elasticsearch keystore?

**Comment by synhershko** [comment]
@dadoonet did you try adding `"java.lang.RuntimePermission" "accessDeclaredMembers"` to `plugin-security.policy`?

**Comment by dadoonet** [comment]
@synhershko Yes I did: https://github.com/elastic/elasticsearch/pull/22679/files#diff-71c14186b73ea943acb9c7a187662278R22

**Comment by dadoonet** [comment]
So latest update is failing with some warnings and I have no idea why because I'm not using this enum `SinceVersion.V1_3_0`:

```
:plugins:discovery-azure-arm:compileJava - is not incremental (e.g. outputs have changed, no previous execution, etc.).
warning: unknown enum constant SinceVersion.V1_3_0
warning: unknown enum constant SinceVersion.V1_3_0
warning: unknown enum constant SinceVersion.V1_3_0
warning: unknown enum constant SinceVersion.V1_3_0
error: warnings found and -Werror specified
1 error
4 warnings
```

@rjernst @jasontedor Does this ring a bell to you?

**Comment by jasontedor** [comment]
That is from the Azure SDK for Java. Look at `com.microsoft.azure.management.apigeneration.Beta.SinceVersion`. Are you sure that your dependency versions are in sync?

**Comment by dadoonet** [comment]
@jasontedor I don't understand why something I don't explicitly call is generating a warning.

To update the deps, I did what I described in https://github.com/elastic/elasticsearch/pull/22679/commits/11fc0e49c8354565928d5687e996f44a1dac67f6

So I created a maven `pom.xml` like:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>fr.pilato.test.azure</groupId>
    <artifactId>azure-deps</artifactId>
    <version>1.0-SNAPSHOT</version>

    <dependencies>
        <dependency>
            <groupId>com.microsoft.azure</groupId>
            <artifactId>azure</artifactId>
            <version>1.4.0</version>
        </dependency>
        <dependency>
            <groupId>com.microsoft.azure</groupId>
            <artifactId>azure-mgmt-compute</artifactId>
            <version>1.4.0</version>
        </dependency>
    </dependencies>
</project>
```

And then I ran:

```sh
mvn dependency:tree
```

To get the full list of transitive deps. Which I applied manually in my gradle file. So I'd say that I'm up to date or at least I did everything I could to be up to date.

**Comment by jasontedor** [comment]
@dadoonet It's an annotation on classes that you have a dependency on. Those get processed by the compiler because of compile-time annotation processing.

**Comment by dadoonet** [comment]
@jasontedor Is there a way to ignore that? And if so is it something "safe" to do?

**Comment by jasontedor** [comment]
Why is the dependency that holds the annotation missing? Why aren't you able to add it?

**Comment by dadoonet** [comment]
Ha thanks @jasontedor. Got it now. For whatever reason, I was pulling indeed the wrong annotation dep.
Let see how it goes now.

**Comment by jasontedor** [comment]
You're welcome.

**Comment by dadoonet** [comment]
@elasticmachine retest this please

**Comment by dadoonet** [comment]
This is now to work well from an elasticsearch node:

```
[2018-07-27T15:15:16,608][DEBUG][o.e.d.a.a.AzureArmUnicastHostsProvider] Ignoring machine [base6/10.0.0.4] because of [DEALLOCATED] power status
[2018-07-27T15:15:16,609][DEBUG][o.e.d.a.a.AzureArmUnicastHostsProvider] found networkAddress for [logstash]: [10.0.1.9]
[2018-07-27T15:15:16,609][TRACE][o.e.d.a.a.AzureArmUnicastHostsProvider] adding 10.0.1.9, transport_address 10.0.1.9:9300
[2018-07-27T15:15:16,609][DEBUG][o.e.d.a.a.AzureArmUnicastHostsProvider] found networkAddress for [lsdata-0]: [10.0.1.6]
[2018-07-27T15:15:16,610][TRACE][o.e.d.a.a.AzureArmUnicastHostsProvider] adding 10.0.1.6, transport_address 10.0.1.6:9300
[2018-07-27T15:15:16,610][DEBUG][o.e.d.a.a.AzureArmUnicastHostsProvider] found networkAddress for [lsdata-1]: [10.0.1.7]
[2018-07-27T15:15:16,610][TRACE][o.e.d.a.a.AzureArmUnicastHostsProvider] adding 10.0.1.7, transport_address 10.0.1.7:9300
[2018-07-27T15:15:16,610][DEBUG][o.e.d.a.a.AzureArmUnicastHostsProvider] found networkAddress for [lsdata-2]: [10.0.1.8]
[2018-07-27T15:15:16,611][TRACE][o.e.d.a.a.AzureArmUnicastHostsProvider] adding 10.0.1.8, transport_address 10.0.1.8:9300
[2018-07-27T15:15:16,611][DEBUG][o.e.d.a.a.AzureArmUnicastHostsProvider] found networkAddress for [lskibana]: [10.0.1.5]
[2018-07-27T15:15:16,611][TRACE][o.e.d.a.a.AzureArmUnicastHostsProvider] adding 10.0.1.5, transport_address 10.0.1.5:9300
[2018-07-27T15:15:16,611][DEBUG][o.e.d.a.a.AzureArmUnicastHostsProvider] found networkAddress for [dpi-arm-test]: [10.0.2.4]
[2018-07-27T15:15:16,612][TRACE][o.e.d.a.a.AzureArmUnicastHostsProvider] adding 10.0.2.4, transport_address 10.0.2.4:9300
[2018-07-27T15:15:16,612][DEBUG][o.e.d.a.a.AzureArmUnicastHostsProvider] 6 hosts(s) added
```

Not sure why the build is not working well yet.
Let's figure this out when the review is done.

In the meantime, someone from @elastic/microsoft team would like to test this new plugin?

**Comment by dadoonet** [comment]
@elasticmachine retest this please

**Comment by dadoonet** [comment]
I don't know why `org.elasticsearch.index.engine.InternalEngineTests.testSeqNoAndCheckpoints` is failing. Does not seem related to my change.

Anyway, someone from the @elastic/es-distributed team would like to review?

**Comment by ywelsch** [comment]
Merging in latest master should fix this (see https://github.com/elastic/elasticsearch/issues/32430)

**Comment by russcam** [comment]
@dadoonet I'd like to test this with the Azure ARM template; is it possible to build a version for 6.3.1?

**Comment by dadoonet** [comment]
@ywelsch I was sure I did. And actually I did not. 😄 

I think I have a trickier thing to solve now.
When I run this locally with JDK10, it passes.

```
./gradlew :plugins:discovery-azure-arm:thirdPartyAudit
```

But with JDK8, this is failing with:

```
08:45:41 Execution failed for task ':plugins:discovery-azure-arm:thirdPartyAudit'.
08:45:41 > Invalid exclusions, nothing is wrong with these classes: [javax/activation/ActivationDataFlavor.class, javax/activation/DataContentHandler.class, javax/activation/DataHandler.class, javax/activation/DataSource.class, javax/activation/FileDataSource.class, javax/activation/FileTypeMap.class]
```

(Source: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request/14717/console)

Do you know how I can make this check optional depending on the JVM version? Unless that's a bad idea and that should be solved in another way?

**Comment by dadoonet** [comment]
Thanks @ywelsch that worked! Should I rebase and squash before anyone review it?

**Comment by ofer-velich** [comment]
wondering when discovery-azure plugin should be available?
what should I do for discovery if I'm running elasticsearch on azure arm instances in the meantime?
tnx

**Comment by ywelsch** [comment]
We discussed this PR during the distributed sync and, while generally useful, think that we currently cannot take this one on, in particular give it the proper testing, support and maintenance that we provide for the other plugins in the ES repo. This might be a better fit as a community plugin for now.

**Comment by dadoonet** [comment]
Can we then remove this part of the documentation? https://www.elastic.co/guide/en/elasticsearch/plugins/current/discovery-azure-classic.html

image

Or fix its content saying that this plugin will be removed in the future?

**Comment by ywelsch** [comment]
sure, thank you for pointing this out.

**Pull request closed without merging**